MPU9250 runs on RPi2/Navio+

bo-rc opened the pull request:
Implement MPU9250 IMU sensor accessing functions via Linux SPI interface for RPi2/Navio+.

### Compilation and tests:

**Native compilation on RPi2**: `make rpi2_native`, run tests

**Cross compilation on Ubuntu**: `make rpi2_cross`, upload test binaries to RPi2 and run tests

### Questions/Concerns:

- ~~why we need to use `m_dev_instance_path` in this [`open()` function](https://github.com/PX4/DriverFramework/blob/a11450bbadc599a1a4478aa5e7e60675a6c07275/framework/include/SPIDevObj.hpp#L93)? i have to change it to `m_dev_path` to make RPi2 work.~~ [This line](https://github.com/PX4/DriverFramework/blob/a11450bbadc599a1a4478aa5e7e60675a6c07275/drivers/mpu9250/MPU9250.cpp#L219) is buggy. See #33 
- ~~i'm not sure if i should take care of this block: [`TODO: add ifdef for endianness`](https://github.com/PX4/DriverFramework/blob/a11450bbadc599a1a4478aa5e7e60675a6c07275/drivers/mpu9250/MPU9250.cpp#L311), and how?~~: should swap for rpi2.
- many `#ifdef` preprocessor directives in the code already.
- it seems snapdragon can change [loopback mode](https://github.com/PX4/DriverFramework/blob/a11450bbadc599a1a4478aa5e7e60675a6c07275/framework/src/SPIDevObj.cpp#L234) by software? i didn't implement this function for RPi2.
- ~~I am not sure if my method to setup the [SPI speed divisor](https://github.com/bo-rc/DriverFramework/blob/4218b54146deb12e5f3bd02e0c21359d2c66ac8b/framework/src/SPIDevObj.cpp#L390) is correct. Please help me.~~: use linux `ioctl()` to set rpi2 SPI speed.

Major changes are in file `framework/src/SPIDevObj.cpp`

Thank you!!

mhkabir commented:
@bo-rc I'd love to help here, but looks like that diff has *massive* formatting changes. Could you please revert those in order for us to better understand your changes?

bo-rc commented:
@mhkabir Thanks man! i'm still learning, sorry for that. Can you take a look if it is hopefully better now?

LorenzMeier commented:
Code style seems ok now. @julianoes We will not want to start using #ifdefs. The platform differences should be handled properly in C++.

bo-rc commented:
Hi @LorenzMeier, speaking of code style, i had run astyle tool of the Firmware (`fix_code_style.sh`). but it generated massive white-space inconsistency with respect to the DriverFramework codebase. Did i use it incorrectly?

julianoes commented on framework/src/SPIDevObj.cpp:
```diff
@@ -86,6 +95,36 @@ int SPIDevObj::readReg(DevHandle &h, uint8_t address, uint8_t &val)
+    return 0;
+#endif // __RPI2
+
 #ifdef __QURT
```
I think we should make this ```#elif defined(__QURT)``` to avoid having both by accident.

bo-rc replied:
Good! Thanks!

julianoes commented on framework/src/SPIDevObj.cpp:
```diff
@@ -256,6 +385,14 @@ int SPIDevObj::setBusFrequency(DevHandle &h, SPI_FREQUENCY freq_hz)
+    int32_t samplerate_div = (int32_t)((SPI_FREQUENCY_20MHZ / freq_hz) + 0.5f) - 1;
+#define MPUREG_SMPLRT_DIV 0x19
+    DF_LOG_DEBUG("samplereate_div = %d", samplerate_div);
+    return _writeReg(MPUREG_SMPLRT_DIV, samplerate_div);
```
I think this is wrong. We want to change the bus frequency of the SPI bus itself. This has nothing to do with the sensor and its samplerate as such.

bo-rc replied:
True! i think i need to do an `ioctl()` call instead.

julianoes commented on framework/src/SPIDevObj.cpp:
```diff
@@ -154,7 +193,36 @@ int SPIDevObj::writeRegVerified(DevHandle &h, uint8_t address, uint8_t val)
+    return 0;
+#else
+
+    uint8_t write_buffer[2];
```
Indenting here seems wrong. Probably best to just run the formatter.

bo-rc replied:
@julianoes can you point to me which formatter i need to run? I used to run astyle tool of the Firmware (`fix_code_style.sh`, default setting parameters in Firmware). but it generated worse indentations.

julianoes commented on framework/src/SPIDevObj.cpp:
```diff
@@ -201,6 +279,43 @@ int SPIDevObj::bulkRead(DevHandle &h, uint8_t address, uint8_t* out_buffer, int
+    /* implement sensor interface via rpi2 spi */
+    int transfer_bytes = 1 + length; // first byte is address
+    uint8_t write_buffer[255] = {0}; // automatic write buffer
+    uint8_t read_buffer[255] = {0}; // automatic read buffer
```
The length of the buffers seems random. What about adapting them to length?

julianoes commented:
Nice additions @bo-rc thank you. I'll look into #33 next.
It would be great if you could rebase this branch on current DriverFramework master. This should also get rid of the code style errors that you're seeing.

bo-rc commented:
Hi @julianoes thank you very much for the review! i revised the code accordingly and the changes are the following:

1. implement setting SPI speed on RPI2 using `ioctl()` call.
2. replace `#ifdef` blocks with `#if define()` blocks to better handle preprocessing code for different architectures.
3. use automatic stack allocation to fit the buffer size requested by user for bulk read MPU9250 registers.
4. some indentation fixes.
5. this current PR is based on the current DriverFramework master

Please have a look and let me know. Thank you again!!

bo-rc commented on framework/src/SPIDevObj.cpp:
```diff
@@ -137,8 +174,8 @@ int SPIDevObj::writeRegVerified(DevHandle &h, uint8_t address, uint8_t val)
-                                --retries;
-                                continue;
+				--retries;
+				continue;
```
I tried to fix this indentation problem but i failed no matter what.

julianoes commented:
@bo-rc: did you cross compile this or only natively on the RPi2? I'm currently trying to cross-compile but presumably using the wrong toolchain.

```
-bash: ./df_imu_test: cannot execute binary file
```

bo-rc commented:
@julianoes Hi Julian, i used to build the code natively on the RPi2. Inspired by your comment, I added a target (`make rpi2`) to cross-compile the code on a Linux computer.  Let me know. Thanks!!

mhkabir commented:
@bo-rc @julianoes 

@blankered would like to assist you on Navio+ driver integration in DF. Please let him know how he can best contribute, and avoid parallel efforts.

bo-rc commented:
@mhkabir @julianoes @blankered 

I am now working on Mag sensor interface for rpi2. GPS sensor is my next task. I agree with @mhkabir , we need to find an organized way to push this forward. Let me know. Thanks!

LorenzMeier commented:
@julianoes Can you try to mainline this today? Let me know if I can help.

julianoes commented:
@bo-rc Thanks for the rpi2 target, the cross compilation seemed to work.
However, the MPU9250 test doesn't seem to be happy with the sensor:
```
./df_imu_test 
0 Framework::initialize
599 HRTWorkQueue::initialize
1107 Created HRTWorkQueue
1252 m_instance = 0x1192008
1410 Calling pthread_mutex_init
1523 pthread_mutex_init success
1719 setRealtimeSched success
2084 HRTWorkQueue::process_trampoline
2101 pthread_create success
2406 Calling DevMgr::initialize
2870 Calling WorkMgr::initialize
3119 Run it
3650 Added driver 0x7ed11578 /dev/imu0
4025 SPI speed set to 1MHz.
4625 MPU9250 sensor WHOAMI wrong: 0x40, should be: 0x71
```

bo-rc commented:
@julianoes thanks for testing! If you are running the Linux kernel described on [this page](https://github.com/PX4/Firmware/issues/3509), you don't need to do any other setting on the RPi2 to make SPI work.


1. Can you reboot your rpi2 board and try it again? 

2. How many times you got wrong WHOAMI values? Can you test it more times to see if a wrong whoiami reading is persistent?

I never get this complaint in my tests. but it could be the case that your SPI chip enable happened to be active for the other spi device, which is supposed to be the Ublox GPS but we haven't implemented it yet.

mhkabir commented:
And make sure the Pi isn't overclocked, and has sufficient power available to it.

julianoes commented:
@bo-rc, @mhkabir:
1. I do have the Linux kernel from that page.
2. I rebooted and tried again.
3. The WHOAMI is always wrong: 0x0, 0x70, 0x40, 0x31, ... .
4. I didn't set any overclocking setting.
5. I tried with a beefy power supply, no change.

bo-rc commented:
@julianoes :
> 2. I rebooted and tried again.

Can you try `sudo shutdown -now` and power off the board for 15min, then try again?
I read some reports from Navio's forum the capacitors of the board might be able to last for over 10min.

> 3. The WHOAMI is always wrong: 0x0, 0x70, 0x40, 0x31, ... 

Do you get this every time you launch the test or after each rebooting but the value is persistent between rebooting?
If you just ignore the WHOAMI test (say, comment out the code), would you be able to get a reading from the sensors?

> 4. I didn't set any overclocking setting.

can you do `sudo /boot/config.txt` and post the content?


@mhkabir can you also test? i tested it successfully but we have to hunt this down. It is also possible that the hardware is faulty. @julianoes can you share your binary? i want to try it on my board. Thanks!

julianoes commented:
@bo-rc:
Thanks for the advice.

> Can you try sudo shutdown -now and power off the board for 15min, then try again?
I read some reports from Navio's forum the capacitors of the board might be able to last for over 10min.

Didn't change the situation.

> Do you get this every time you launch the test or after each rebooting but the value is persistent between rebooting?

I get it every time, and the value seems random.

> If you just ignore the WHOAMI test (say, comment out the code), would you be able to get a reading from the sensors?

Interesting, the temperature seems to be sane, and changing, so presumably the writes fail (so it's not configured but the reads work anyway):

```
3536475 count: 878
3536612 IMU: accel: [-153.22, 0.16, 0.00] m/s^2
3536751      gyro:  [0.06, -1234.25, 0.06] rad/s
3536955      temp:  35.71 C
```

> can you do cat /boot/config.txt and post the content?

Here: http://pastebin.com/M1HHkiqJ

bo-rc commented:
@julianoes your config.txt is exactly the same as mine.

Can you try my binary here: https://github.com/bo-rc/Scr/blob/master/df_imu_test
right-click, save link as ... to RPi2

I want to try your binary on my board too. Thanks!

LorenzMeier commented:
Julian has a Navio 2 - did you factor that in?

bo-rc commented:
@LorenzMeier Oh thanks!! @julianoes are you using a Navio 2? which has a redundant IMU and should be treated differently. This PR is for Navio+ as titled. 

I don't have a Navio 2, so i cannot test it.

julianoes commented:
Aha :)

bo-rc commented:
@julianoes can you try to use this kernel image instead:
http://docs.emlid.com/navio2/Navio-APM/configuring-raspberry-pi/

and try the driver test again.

mhkabir commented:
Aha as well :)

I have a NAVIO 2 here. Let me set that up, and get started with the
development.

As Bo said, the Navio2 has multiple things on SPI.

@julianoes <https://github.com/julianoes> can you try to use this kernel
image instead:
http://docs.emlid.com/navio2/Navio-APM/configuring-raspberry-pi/

and try the driver test again.

—
Reply to this email directly or view it on GitHub
<https://github.com/PX4/DriverFramework/pull/32#issuecomment-189368800>.

blankered commented:
I am working with the Navio+ with jeesie image and rtk from emlid. I tried testing the mpu9250. This was my output.

`
pi@raspberrypi:~/DriverFramework/build_linux/drivers/mpu9250/test $ ./df_imu_test
0 Framework::initialize
288 HRTWorkQueue::initialize
553 Created HRTWorkQueue
751 m_instance = 0x654008
813 Calling pthread_mutex_init
913 pthread_mutex_init success
1070 setRealtimeSched success
1426 pthread_create success
1538 Calling DevMgr::initialize
1620 Calling WorkMgr::initialize
1713 Run it
1808 Added driver 0x7e954290 /dev/imu0
1965 Unable to open the device path: /dev/spidev0.0
1445 HRTWorkQueue::process_trampoline
`

mhkabir commented:
Try sudoing mainapp for a quick hack trial.

blankered commented:
I checked the Navio+ drivers. The example app has the imu data loaded on /dev/spidev0.1 not /dev/spidev0.0 as listed in the DriverFrameworks Imu app. Is this an issue?

bo-rc commented:
Hi @blankered thanks for testing!

> `1965 Unable to open the device path: /dev/spidev0.0`

Sorry, when i created the cross-compilation target `rpi2`, i forgot the native one. So to clarify, i added a new target called `rpi2_native` and renamed the cross-compilation target to `rpi2_cross`.

Can you update the code and test again? Thanks!

bo-rc commented:
On Fri, Feb 26, 2016 at 2:24 PM, blankered <notifications@github.com> wrote:

> I checked the Navio+ drivers. The example app has the imu data loaded on
> /dev/spidev0.1 not /dev/spidev0.0 as listed in the DriverFrameworks Imu
> app. Is this an issue?
>
The code should pick up /dev/spidev0.1 when compiled on RPI2 (when the
macro __RPI2 is defined).
My very recent push should fix this. The code should pick up the macro now.
Speaking of preprocessor #define's, as Lorenz said before, we will use C++
and build system to take care of architecture differences.

> —
> Reply to this email directly or view it on GitHub
> <https://github.com/PX4/DriverFramework/pull/32#issuecomment-189468134>.
>

blankered commented:
@bo-rc  I tested your df_imu_test compiled app. Works great! Also, I have been compiling with rpi2 gcc, g++, cpp 4.9 compilers on the rpi2 not the toolchain. I haven't had any issues, let me know if there are any problems with this approach.

bo-rc commented:
@blankered thank you! Your approach (native compilation, as i can see) is fine!

LorenzMeier commented:
@mhkabir @julianoes Could you please do another code review and provide feedback? I'd like to get this in as soon as possible, now that we validated QuRT.

julianoes commented:
@bo-rc thanks for your contributions here, I gave it another quick review and it looks good. I've rebased this in #42 and I'll close this PR.

Unfortunately, I currently don't have time to look into support for Navio2, I made an issue (#44) to track it. If anyone has the hardware available and free cycles, that would be great.
